File: akka/src/main/scala/se/kth/benchmarks/akka/bench/Fibonacci.scala
package se.kth.benchmarks.akka.bench

import se.kth.benchmarks.akka.ActorSystemProvider
import se.kth.benchmarks.Benchmark
import kompics.benchmarks.benchmarks.FibonacciRequest
import akka.actor.{Actor, ActorLogging, ActorRef, ActorSystem, PoisonPill, Props}
import scala.concurrent.Await
import scala.concurrent.duration._
import scala.util.Try
import java.util.concurrent.CountDownLatch
import com.typesafe.scalalogging.StrictLogging

object Fibonacci extends Benchmark {
  override type Conf = FibonacciRequest;

  override def msgToConf(msg: scalapb.GeneratedMessage): Try[Conf] = {
    Try(msg.asInstanceOf[FibonacciRequest])
  };
  override def newInstance(): Instance = new FibonacciI;

  class FibonacciI extends Instance with StrictLogging {

    private var fibNumber = -1;
    private var system: ActorSystem = null;
    private var fib: ActorRef = null;
    private var latch: CountDownLatch = null;

    override def setup(c: Conf): Unit = {
      logger.info("Setting up Instance");
      this.fibNumber = c.fibNumber;
      this.system = ActorSystemProvider.newActorSystem(name = "fibonacci");
    }
    override def prepareIteration(): Unit = {
      assert(this.system != null);
      logger.debug("Preparing iteration");
      this.latch = new CountDownLatch(1);
      this.fib = system.actorOf(Props(new FibonacciActor(Right(latch))));
    }
    override def runIteration(): Unit = {
      assert(this.fibNumber > 0);
      assert(this.latch != null);
      assert(this.fib != null);
      this.fib ! FibonacciMsg.Request(this.fibNumber);
      latch.await();
    }
    override def cleanupIteration(lastIteration: Boolean, execTimeMillis: Double): Unit = {
      logger.debug("Cleaning up iteration");
      if (this.latch != null) {
        this.latch = null;
      }
      if (this.fib != null) {
        // system.stop(this.fib); kills itself
        this.fib = null;
      }
      if (lastIteration) {
        val f = system.terminate();
        Await.ready(f, 5.second);
        system = null;
        logger.info("Cleaned up Instance");
      }
    }
  }

  sealed trait FibonacciMsg;
  object FibonacciMsg {
    final case class Request(n: Int) extends FibonacciMsg;
    final case class Response(value: Long) extends FibonacciMsg;
  }

  class FibonacciActor(val reportTo: Either[ActorRef, CountDownLatch]) extends Actor with ActorLogging {
    import FibonacciMsg._;

    private var result = 0L;
    private var numResponses = 0;

    override def receive = {
      case Request(n) => {
        log.debug(s"Got Request with n=$n");
        if (n <= 2) {
          sendResult(1L);
        } else {
          val f1 = context.actorOf(Props(new FibonacciActor(Left(self))));
          f1 ! Request(n - 1);
          val f2 = context.actorOf(Props(new FibonacciActor(Left(self))));
          f2 ! Request(n - 2);
        }
      }
      case Response(value) => {
        log.debug(s"Got Response with value=$value");
        this.numResponses += 1;
        this.result += value;

        if (this.numResponses == 2) {
          sendResult(this.result)
        }
      }
    }

    private def sendResult(value: Long): Unit = {
      reportTo match {
        case Left(ref) => ref ! Response(value)
        case Right(latch) => {
          latch.countDown();
          log.info(s"Final value was $value");
        }
      }
      context.stop(self);
    }
  }
}

File: riker/src/bench/fibonacci.rs
use super::*;

use crate::riker_system_provider::*;
use benchmark_suite_shared::kompics_benchmarks::benchmarks::FibonacciRequest;
use log::debug;
use riker::actors::*;
use std::sync::Arc;
use synchronoise::CountdownEvent;

#[derive(Default)]
pub struct Fibonacci;

impl Benchmark for Fibonacci {
    type Conf = FibonacciRequest;
    type Instance = FibonacciI;

    fn msg_to_conf(msg: Box<dyn (::protobuf::Message)>) -> Result<Self::Conf, BenchmarkError> {
        downcast_msg!(msg; FibonacciRequest)
    }

    fn new_instance() -> Self::Instance {
        FibonacciI::new()
    }

    const LABEL: &'static str = "Fibonacci";
}

pub struct FibonacciI {
    fib_number: Option<u32>,
    system: Option<RikerSystem>,
    fib: Option<ActorRef<FibonacciMsg>>,
    latch: Option<Arc<CountdownEvent>>,
}

impl FibonacciI {
    fn new() -> FibonacciI {
        FibonacciI {
            fib_number: None,
            system: None,
            fib: None,
            latch: None,
        }
    }
}

impl BenchmarkInstance for FibonacciI {
    type Conf = FibonacciRequest;

    fn setup(&mut self, c: &Self::Conf) -> () {
        self.fib_number = Some(c.fib_number);
        let system = RikerSystem::new("fibonacci", num_cpus::get()).expect("System");
        self.system = Some(system);
    }

    fn prepare_iteration(&mut self) -> () {
        match self.system {
            Some(ref system) => {
                let latch = Arc::new(CountdownEvent::new(1));
                let fib = system
                    .start(
                        FibonacciActor::props(ResultTarget::Latch(latch.clone())),
                        "fib",
                    )
                    .expect("Should start FibonacciActor!");

                self.fib = Some(fib);
                self.latch = Some(latch);
            }
            None => unimplemented!(),
        }
    }

    fn run_iteration(&mut self) -> () {
        match self.fib_number {
            Some(fib_number) => match self.fib {
                Some(ref fib) => {
                    let latch = self.latch.take().unwrap();

                    fib.tell(FibonacciMsg::request(fib_number), None);
                    latch.wait();
                }
                None => unimplemented!(),
            },
            None => unimplemented!(),
        }
    }

    fn cleanup_iteration(&mut self, last_iteration: bool, _exec_time_millis: f64) -> () {
        let _ = self.fib.take(); // kills itself, so just drop

        if last_iteration {
            let system = self.system.take().unwrap();
            system
                .shutdown()
                .wait()
                .expect("Rikerdidn't shut down properly");
            self.fib_number = None;
        }
    }
}

#[derive(Clone)]
enum ResultTarget {
    Latch(Arc<CountdownEvent>),
    Parent(ActorRef<FibonacciMsg>),
}

#[derive(Debug, Clone)]
enum FibonacciMsg {
    Request { n: u32 },
    Response { value: u64 },
}
impl FibonacciMsg {
    fn request(n: u32) -> FibonacciMsg {
        FibonacciMsg::Request { n }
    }

    fn response(value: u64) -> FibonacciMsg {
        FibonacciMsg::Response { value }
    }

    const RESPONSE_ONE: FibonacciMsg = FibonacciMsg::Response { value: 1u64 };
}

struct FibonacciActor {
    report_to: ResultTarget,
    result: u64,
    num_responses: u8,
}
impl FibonacciActor {
    fn with(report_to: ResultTarget) -> FibonacciActor {
        FibonacciActor {
            report_to,
            result: 0u64,
            num_responses: 0u8,
        }
    }

    fn props(report_to: ResultTarget) -> BoxActorProd<FibonacciActor> {
        Props::new_from(move || FibonacciActor::with(report_to.clone()))
    }

    fn send_result(&mut self, response: FibonacciMsg, ctx: &Context<FibonacciMsg>) {
        match self.report_to {
            ResultTarget::Parent(ref parent_ref) => {
                parent_ref.tell(response, None);
            }
            ResultTarget::Latch(ref latch) => {
                latch.decrement().expect("Should decrement!");
            }
        }
        ctx.stop(ctx.myself());
    }
}

impl Actor for FibonacciActor {
    type Msg = FibonacciMsg;

    fn recv(&mut self, ctx: &Context<Self::Msg>, msg: Self::Msg, _sender: Sender) {
        match msg {
            FibonacciMsg::Request { n } => {
                debug!("Got Request n={}", n);
                if n <= 2u32 {
                    self.send_result(FibonacciMsg::RESPONSE_ONE, ctx);
                } else {
                    let self_ref = ctx.myself();
                    let f1 = ctx
                        .actor_of_props(
                            "fib1",
                            FibonacciActor::props(ResultTarget::Parent(self_ref.clone())),
                        )
                        .expect("Fibonacci Child");
                    f1.tell(FibonacciMsg::request(n - 1u32), None);
                    let f2 = ctx
                        .actor_of_props(
                            "fib2",
                            FibonacciActor::props(ResultTarget::Parent(self_ref.clone())),
                        )
                        .expect("Fibonacci Child");
                    f2.tell(FibonacciMsg::request(n - 2u32), None);
                }
            }
            FibonacciMsg::Response { value } => {
                debug!("Got Response value={}", value);
                self.num_responses += 1u8;
                self.result += value;

                if self.num_responses == 2u8 {
                    self.send_result(FibonacciMsg::response(self.result), ctx);
                }
            }
        }
    }
}
